crypto/fipsmodule/ssh/kdf.c

```diff
@@ -0,0 +1,105 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+// Based on OpenSSL 3.x SSHKDF().
+
+#include <openssl/digest.h>
+#include <openssl/mem.h>
+
+#include "../service_indicator/internal.h"
+#include "../../internal.h"
+
+int SSHKDF(const EVP_MD *evp_md,
+           const uint8_t *key, size_t key_len,
+           const uint8_t *xcghash, size_t xcghash_len,
+           const uint8_t *session_id, size_t session_id_len,
+           char type,
+           uint8_t *out, size_t out_len)
+{
+  EVP_MD_CTX *md = NULL;
+  uint8_t digest[EVP_MAX_MD_SIZE];
+  unsigned int digest_size = 0;
+  size_t cursize = 0;
+  int ret = 0;
+
+  FIPS_service_indicator_lock_state();
+
+  md = EVP_MD_CTX_new();
+  if (md == NULL) {
+    return 0;
+  }
+
+  if (!EVP_DigestInit_ex(md, evp_md, NULL)) {
```

Comment:
Are branches corresponding to `EVP_DigestXYZ` APIs failures have tests?

Reply:
They should, in the digest unit tests.

Reply:
Is there any unit tests for `SSHKDF` that would cause `!EVP_DIgestInit_ex(md, evp_md, NULL)` fail and cover that branch in this file?

Reply:
The unit tests in AWS-LC seem to cover the API being tested, not the APIs they're calling; the PBKDF and HKDF unit tests (for example) don't test the underlying HMAC calls.

If we want to improve code coverage by triggering all (or any? more?) branches, we'll have a lot of additional tests to add. I'm not saying that would be a bad thing, just that we don't currently have it.